File: LeetCode/52.N-Queens_II/Mario/52.N-Queens_II.py
#!/usr/bin/env python3


class Solution:
    def totalNQueens(self, n: int) -> int:
        res = 0
        cols, diag1, diag2 = set(), set(), set()

        def backtracking(row):
            nonlocal res
            if row == n:
                res += 1
                return

            for c in range(n):
                if c in cols or (row + c) in diag1 or (row - c) in diag2:
                    continue

                cols.add(c)
                diag1.add(row + c)
                diag2.add(row - c)

                backtracking(row + 1)

                cols.remove(c)
                diag1.remove(row + c)
                diag2.remove(row - c)

        backtracking(0)

        return res

File: LeetCode/52.N-Queens_II/52.N-Queens_II.java
class Solution {
    int res = 0;

    public int totalNQueens(int n) {
        Set<Integer> col = new HashSet<>();
        Set<Integer> posDiagonal = new HashSet<>();
        Set<Integer> negDiagonal = new HashSet<>();
        backtracking(n, 0, col, posDiagonal, negDiagonal);
        return res;
    }

    private void backtracking(int n, int r, Set<Integer> col, Set<Integer> posDiagonal, Set<Integer> negDiagonal) {
        if (r == n) {
            res += 1;
            return;
        }

        for (int c = 0; c < n; c++) {
            if (col.contains(c) || posDiagonal.contains(r + c) || negDiagonal.contains(r - c)) {
                continue;
            }
            col.add(c);
            posDiagonal.add(r + c);
            negDiagonal.add(r - c);
            backtracking(n, r + 1, col, posDiagonal, negDiagonal);
            col.remove(c);
            posDiagonal.remove(r + c);
            negDiagonal.remove(r - c);
        }
    }
}